### PR TITLE
feat: opt-in MeterId output in pricing scripts (#530)

### PR DESCRIPTION
**Pull request by ahmadabdalla**
## Summary

Proof of concept for [#530](https://github.com/ahmadabdalla/azure-cost-calculator/issues/530) — adds opt-in `--include-meter-id` / `-IncludeMeterId` flag to all 4 pricing scripts. When set, `MeterId` (GUID) appears in Json/Compact output for cross-referencing with Azure Cost Management, invoices, and billing exports.

- **Zero default overhead** — MeterId is captured internally but excluded from output unless the flag is set
- **Exclusion-list pattern** — `$excludeProps` (PS) / `$exclude_fields` (Bash) built once after arg parsing, applied once before output. Adding future optional fields = one new param + one line in the exclusion list, zero changes to output sections
- **Table/Summary unaffected** — GUIDs never appear in columnar/text formats
- **Bug fix** — `Explore-AzurePricing.ps1` now returns a JSON array for single results (`-AsArray`), matching the Bash equivalent

## Remaining work

- [x] A/B test: run two Opus 4.6 sessions against `data-analytics-platform.md` example (with MeterId vs without) to measure UX and accuracy — see results below
- [ ] Add bats + Pester tests for `--include-meter-id` flag presence/absence and null-safety
- [ ] Add keyword trigger in `SKILL.md` (detect "meter ID", "meterId", "reconciliation", etc.)
- [ ] Update `references/workflow.md` to document the new parameter
- [ ] Consider `-MeterId` input parameter for direct GUID lookups (OData filter `meterId eq '...'`)
- [ ] Validate null-safety for `hasMeters: false` / `pricingRegion: api-unavailable` services

## A/B Test — `data-analytics-platform.md` (Opus 4.6)

Two independent Opus 4.6 sessions estimated the full **Regulatory & Portfolio Analytics Platform** (15 services, 31 line items, uksouth/GBP, 1-Year RI + AHUB) — one with `--include-meter-id` and one without.

### Price convergence

| Category | Without MeterId | With MeterId | Delta |
|---|---|---|---|
| IoT | £196.07 | £196.07 | £0.00 |
| Analytics | £9,994.43 | £9,994.41 | −£0.02 |
| Compute | £3,955.92 | £3,955.92 | £0.00 |
| Storage | £3,695.08 | £3,695.15 | +£0.07 |
| Databases | £1,922.18 | £1,932.83 | +£10.65 |
| Networking | £740.39 | £740.35 | −£0.04 |
| Security | £199.08 | £199.08 | £0.00 |
| **Grand Total** | **£20,703.15** | **£20,713.81** | **+£10.66** |

### Findings

- **99.95% price agreement** — the £10.66 delta is not rounding; the MeterId session used the explore script to discover the SQL Database LTR Backup Storage meter (`901630c5-...`, £0.0213/GB), which is distinct from Azure Backup's generic LRS meter (`c512f8b3-...`, £0.0207/GB). The architecture lists both SQL DB LTR and Azure Backup as separate items, so the MeterId session correctly priced both as separate line items.
- **26 unique MeterIds** returned across 31 line items (some meters are reused, e.g., Event Hubs TU across both namespaces, Backup LRS across SQL DB + ADLS storage).
- **No token efficiency regression** — Compact format with MeterId adds only 1 field per result row (10 vs 9 fields). Both sessions consumed similar tool-call counts (~30 API queries each).
- **All unit prices identical** across both sessions — same AHUB logic, same RI division, same USD→GBP conversion factor (0.7394 via B2s anchor).

### MeterId sample (from the with-MeterId session)

| Service | Meter | MeterId |
|---|---|---|
| Event Hubs TU | Standard Throughput Unit | `62d94a65-9300-48a6-8c15-0e70fc41eb44` |
| Databricks DBU | Premium Jobs Compute DBU | `8ee07b23-82ae-430e-a417-a050463ff4ee` |
| SQL DB BC Compute | vCore | `6691a172-a325-4470-881e-4354c34e2caf` |
| Synapse DW100c | 100 DWUs | `83778534-91d7-4960-84ac-80307a16405b` |
| Azure Firewall | Standard Deployment | `e90604a4-719b-454b-879d-b85523882c92` |
| Private Link | Standard Private Endpoint | `e6ab7238-e433-4fe0-a2b2-2b2564df2cdb` |

## Test plan

- [x] All 153 bats tests pass
- [x] Live tested all output formats (Json, Compact, Table, Summary) × both runtimes (Bash, PowerShell) × with/without flag
- [x] Side-by-side worktree comparison: default output identical to dev (zero regression)
- [x] Verified MeterId GUID values match across Bash and PowerShell for same query
- [x] Full architecture A/B test (`data-analytics-platform.md`, 31 line items) — 99.95% price convergence, 26 unique MeterIds verified

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an optional IncludeMeterId / --include-meter-id flag to Azure pricing tools; MeterId is omitted by default and included when requested in JSON and Compact outputs.

* **Documentation**
  * Updated help text, usage, and reference docs to document the new option and its effect on Compact field count.

* **Tests**
  * New unit tests cover include/exclude behavior and null-MeterId scenarios across formats.

### CODE REVIEW

File: skills/azure-cost-calculator/scripts/Get-AzurePricing.ps1 [hunk 1]
```diff
@@ -55,11 +55,18 @@ param(
 
     [Parameter()]
     [ValidateSet('Table', 'Json', 'Summary', 'Compact')]
-    [string]$OutputFormat = 'Json'
+    [string]$OutputFormat = 'Json',
+
+    [Parameter()]
+    [switch]$IncludeMeterId
 )
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply skills/azure-cost-calculator/scripts/Get-AzurePricing.ps1]: Thanks for the scaffold — we've implemented comprehensive tests that go beyond this:

**Pester (Get-AzurePricing.Tests.ps1)** — 6 new contexts:
- Default Json excludes MeterId
- `-IncludeMeterId` includes MeterId in Json with correct GUID value
- Default Compact: 9 fields, no MeterId
- `-IncludeMeterId` Compact: 10 fields with MeterId
- Null meterId from API: property present with null value

**Pester (Explore-AzurePricing.Tests.ps1)** — 3 new contexts:
- Default excludes MeterId
- `-IncludeMeterId` includes MeterId with correct value
- Null meterId null-safety

**bats (get-azure-pricing.bats)** — 5 new tests:
- Default Json excludes MeterId
- `--include-meter-id` includes MeterId in Json
- Default Compact: 9 fields, no MeterId
- `--include-meter-id` Compact: 10 fields with MeterId
- Null meterId produces null when included

**bats (explore-azure-pricing.bats)** — 3 new tests:
- Default Json excludes MeterId
- `--include-meter-id` includes MeterId in Json
- Null meterId null-safety

All mock data uses proper `meterId` GUID values via `Invoke-RestMethod` mocks (Pester) and `create_curl_mock` (bats) — no live API calls. All tests passing.

**Comment by ahmadabdalla** [comment]
## Test Scenario — `event-driven-serverless.md` (Opus 4.6, with MeterId)

Full architecture estimate of the **Trade Settlement & AML Event Processing** example (16 services, 20 line items, australiaeast/AUD, 1-Year RI, no AHUB) using `--include-meter-id` and `--output-format Compact` throughout.

### Configuration

- **Example file:** `references/examples/event-driven-serverless.md`
- **Region:** australiaeast | **Currency:** AUD
- **Commitment:** 1-Year Reserved (Cosmos DB, SQL Database)
- **Hybrid Benefit:** Not applied
- **Sentinel pricing model:** Simplified (App Insights 25 GB absorbed into Sentinel 60 GB total)
- **USD→AUD conversion factor:** 1.4072 (B2s anchor, for Private Link + Private DNS)

### MeterId coverage — 17 unique MeterIds across 20 line items

| Category | Service | Meter | MeterId |
|---|---|---|---|
| IoT | Event Grid | Standard Operations | `d207c4bb-d8a3-4f48-bc91-4b324a1399c2` |
| Integration | Service Bus | Premium Messaging Unit | `176d7148-0033-48c7-8b18-7af4a9d1dca7` |
| Compute | Functions Flex | On Demand Execution Time | `3a1d8811-3c16-54ab-8b56-c387b1482e35` |
| Compute | Functions Flex | On Demand Total Executions | `14b7c07d-434e-5b4d-bb1d-67e2be800129` |
| Databases | Cosmos DB | 100 RU/s (1-Year RI) | `65d4ded2-41ae-43a8-bb68-3c200e1ba864` |
| Databases | Cosmos DB | Data Stored | `572f570a-de2f-4407-aef6-f6c8a5e992b3` |
| Databases | Cosmos DB | Cont 30D PITR Continuous Backup | `a8a2fd4f-dd61-4100-9523-4a4d5acf8818` |
| Databases | SQL Database | vCore GP Gen5 (1-Year RI) | `2d6ad12e-49dd-4279-9c6b-e0cd904957ab` |
| Databases | SQL Database | SQL License | `6f344f4b-7c91-4026-915a-a1b597793be1` |
| Databases | SQL Database | GP Data Stored | `875f798e-2ab2-4db8-b5f3-9aa24c247230` |
| Databases | Redis Cache | C2 Cache (Standard) | `7c8f5b20-386c-4078-85de-14d64b3e059d` |
| Storage | Blob Storage | Hot LRS Data Stored | `2f968df1-9a17-4398-bcfb-ad8bf7922321` |
| Storage | Blob Storage | Cool LRS Data Stored | `8405dbf3-ffda-4d3c-85e8-716ec6bd2fbf` |
| Security | Key Vault | Operations (Standard) | `0f824807-2376-435c-95c9-f992b67a07b0` |
| Security | Sentinel | Pay-as-you-go Analysis | `4c6c0fb7-0d91-5007-8b0e-502dae3d5fcc` |
| Networking | Private Link | Standard Private Endpoint | `e6ab7238-e433-4fe0-a2b2-2b2564df2cdb` |
| Networking | Private Link | Data Processed — Ingress | `10caa28b-6479-4852-9eb7-610870cb6417` |
| Networking | Private Link | Data Processed — Egress | `e7f162f6-7cb8-4cea-ad4f-12cdb5dda25b` |
| Networking | Private DNS | Private Zone | `0b608a26-f611-4232-8192-ce81b6b57194` |
| Networking | Private DNS | Private Queries | `46cff9cb-4b15-4af0-bf41-25143827fef8` |
| Networking | Bandwidth | Standard Data Transfer Out | `fe167397-a38d-43c3-9bb3-8e2907e56a41` |

### Cost summary

| Category | Monthly (AUD) |
|---|---|
| IoT | 21.02 |
| Integration | 952.43 |
| Compute | 989.63 |
| Databases | 1,552.59 |
| Storage | 45.05 |
| Security | 527.71 |
| Monitoring | 0.00 |
| Networking | 113.64 |
| **Grand Total** | **4,202.07** |

### Findings

- **17 unique MeterIds** returned across 20 line items — all present and non-null in Compact output with `--include-meter-id`
- **Sub-cent meter handling validated:** Functions On Demand Total Executions returned `UnitPrice: 0.0000` in AUD (sub-cent floor). MeterId (`14b7c07d-...`) was still correctly populated despite zero-display price. USD fallback rate (0.000004/10 exec) used for cost calculation with currency conversion
- **RI + Global region meters work:** Cosmos DB RI (Region: Global) returned MeterId `65d4ded2-...` across all SKU tiers (100 RU/s through 30M RU/s). Same GUID for all tiers — differentiator is `skuName`, not MeterId
- **USD-only services (Private Link, Private DNS):** MeterIds returned correctly from Global/Zone 1 queries. Same PE MeterId (`e6ab7238-...`) as the data-analytics-platform A/B test — confirms cross-session GUID stability
- **Tiered meters:** Event Grid (free + paid tiers) and Bandwidth (free + 5 paid tiers) both returned the same MeterId across all tier rows, with MeterId correctly present on the $0 free-tier rows too
- **Compact format confirmed:** All queries used `--output-format Compact --include-meter-id`, returning 10 fields per row (9 base + MeterId). No regressions in field parsing or JSON structure
- **PITR meters:** Cosmos DB PITR product returned 4 distinct meters with 3 unique MeterIds (7-day free, 30-day backup, 30-day restore, 30-day on-demand) — all correctly populated

### Edge cases exercised

| Edge Case | Result |
|---|---|
| Sub-cent AUD price (Functions executions) | MeterId present, UnitPrice displays 0.0000 |
| Free-tier rows (Event Grid, Bandwidth, Functions) | MeterId present on $0 rows |
| Global region (Cosmos DB RI, Private Link) | MeterId present |
| Empty-region workaround (Private DNS via Zone 1) | MeterId present |
| Multi-tier results (Blob Hot LRS — 3 tiers) | Same MeterId across all tier rows |
| RI results (Cosmos DB — 12 SKU variants) | Same MeterId across all SKU/term variants |

🤖 Generated with [Claude Code](https://claude.ai/code)
